File: reference/macos-swiftui/AirX/Model/Peer.swift
//
//  Peer.swift
//  AnyDropmac
//
//  Created by Hatsune Miku on 2023-06-28.
//

import Foundation

public class Peer: Identifiable {
    let hostName: String
    let host: String
    let port: UInt16
    
    public static let sample = Peer(hostName: "Shijunyi", host: "192.168.0.2", port: 9819)
    
    public var description: String {
        return "\(hostName) (\(host):\(port))"
    }
    
    public init(hostName: String, host: String, port: UInt16) {
        self.hostName = hostName
        self.host = host
        self.port = port
    }
    
    public static func fromUid(uid: Int) -> Peer {
        return Peer(hostName: "UID=\(uid)", host: "AnyDrop Kafka Cluster", port: 0)
    }
    
    /// Peer format: <hostname>@<host>:<port>
    public static func parse(_ s: String) -> Optional<Peer> {
        // Incomplete peer string?
        var peerString = s;
        if !peerString.contains("@") {
            peerString = "<empty>@" + s;
        }
        let part1 = peerString.split(separator: "@")
        guard part1.count == 2 else {
            return .none
        }
        
        let part2 = part1[1].split(separator: ":")
        guard part2.count == 2 else {
            return .none
        }
        
        let hostname = part1[0]
        let host = part2[0]
        let port = UInt16(part2[1])
        return .some(Peer(hostName: String(hostname), host: String(host), port: port!))
    }
}

File: reference/windows-winui/AirX/Model/Peer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AnyDrop.Model
{
    public class Peer
    {
        public string Hostname { get; set; }
        public string IpAddress { get; set; }
        public int Port { get; set; }

        public static readonly Peer Sample = new()
        {
            Hostname = "chance",
            IpAddress = "10.0.0.6",
            Port = 12345,
        };

        /// <summary>
        /// Peer string format: <hostname>@<ip>:<port>
        /// </summary>
        public static Peer Parse(string s)
        {
            // Incomplete peer string?
            if (!s.Contains("@"))
            {
                s = "<empty>@" + s;
            }

            var part1 = s.Split("@");
            var part2 = part1[1].Split(":");
            return new Peer
            {
                Hostname = part1[0],
                IpAddress = part2[0],
                Port = int.Parse(part2[1]),
            };
        }

        public static bool TryParse(string s, out Peer peer)
        {
            try
            {
                peer = Parse(s);
                return true;
            }
            catch (Exception)
            {
                peer = null;
                return false;
            }
        }

        public static Peer FromUid(int uid)
        {
            return new Peer
            {
                Hostname = $"UID={uid}",
                IpAddress = "AnyDrop Kafka Cluster",
                Port = 0,
            };
        }

        public override string ToString()
        {
            return $"{Hostname}@{IpAddress}:{Port}";
        }
    }
}
